compilers/concrete-compiler/compiler/lib/Dialect/FHELinalg/IR/FHELinalgOps.cpp

```diff
@@ -1687,6 +1687,494 @@ void ToUnsignedOp::getCanonicalizationPatterns(
+                  .getT())
+           : (llvm::dyn_cast<ApplyMappedLookupTableEintOp>(
+                  intermediateOperation)
+                  .getT()));
+
+  class Indexer {
+  public:
+    int64_t tableSize;
+    bool isSigned;
+
+    Indexer(int64_t tableSize, bool isSigned)
+        : tableSize{tableSize}, isSigned{isSigned} {}
+
+    virtual ~Indexer() = default;
+
+    int64_t sanitizeIndex(int64_t index) const {
+      if (index < 0) {
+        index += tableSize;
+        if (index < 0) {
+          index = tableSize / 2;
+        }
+      } else if (index >= tableSize) {
+        if (!isSigned) {
+          index = tableSize - 1;
+        } else {
+          index = (tableSize / 2) - 1;
+        }
+      }
+      return index;
+    }
+
+    virtual int64_t get(int64_t index, int64_t position) const = 0;
+  };
+
+  class TluIdexer : public Indexer {
+  public:
+    mlir::Value tableValue;
+    arith::ConstantOp tableOperation;
+    mlir::DenseIntElementsAttr tableContentAttr;
+    std::vector<int64_t> tableContent;
+
+    TluIdexer(int64_t tableSize, bool isSigned, mlir::Value tableValue,
+              arith::ConstantOp tableOperation,
+              mlir::DenseIntElementsAttr tableContentAttr,
+              std::vector<int64_t> tableContent)
+        :
+
+          Indexer{tableSize, isSigned}, tableValue{tableValue},
+          tableOperation{tableOperation}, tableContentAttr{tableContentAttr},
+          tableContent{std::move(tableContent)} {}
+
+    ~TluIdexer() override = default;
+
+    static std::optional<std::unique_ptr<TluIdexer>>
+    create(ApplyLookupTableEintOp operation) {
+
+      auto tableValue = operation.getLut();
+
+      auto tableOperation =
+          llvm::dyn_cast_or_null<arith::ConstantOp>(tableValue.getDefiningOp());
+      if (!tableOperation) {
+        return std::nullopt;
+      }
+
+      auto tableContentAttr =
+          tableOperation.getValueAttr()
+              .dyn_cast_or_null<mlir::DenseIntElementsAttr>();
+      if (!tableContentAttr) {
+        return std::nullopt;
+      }
+
+      auto tableContent = std::vector<int64_t>();
+      for (auto value : tableContentAttr.getValues<int64_t>()) {
+        tableContent.push_back(value);
+      }
+
+      auto inputValue = operation.getT();
+      auto inputType = inputValue.getType()
+                           .cast<RankedTensorType>()
+                           .getElementType()
+                           .dyn_cast<FheIntegerInterface>();
+
+      auto tableSize = 1 << inputType.getWidth();
+      auto isSigned = inputType.isSigned();
+
+      return std::make_unique<TluIdexer>(
+          TluIdexer(tableSize, isSigned, tableValue, tableOperation,
+                    tableContentAttr, tableContent));
+    };
+
+    int64_t get(int64_t index, int64_t position) const override {
+      return tableContent[sanitizeIndex(index)];
+    }
+  };
+
+  class MappedTluIdexer : public Indexer {
```

Comment:
There is a lot of common structure between `TluIndexer` and `MappedTluIndexer`. There is probably room for factoring the differences into specialized templates and collapsing the remainder into a single class template.